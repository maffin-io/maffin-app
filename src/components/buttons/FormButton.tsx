import React from 'react';

import { DataSourceContext } from '@/hooks';
import Modal, { ModalRef } from '@/components/ui/Modal';

export interface FormButtonProps extends Omit<
React.ButtonHTMLAttributes<HTMLButtonElement>,
'children'
> {
  buttonContent: string | React.ReactElement,
  modalTitle: string,
  children: React.ReactElement,
}

export default function FormButton({
  buttonContent,
  modalTitle,
  children,
  className = 'btn btn-primary',
  ...props
}: FormButtonProps): React.JSX.Element {
  const { save } = React.useContext(DataSourceContext);
  const modalRef = React.useRef<ModalRef>(null);

  return (
    <Modal
      ref={modalRef}
      className="modal card"
      triggerContent={buttonContent}
      triggerProps={{ ...props, className }}
      showClose
    >
      <span>{modalTitle}</span>
      <div>
        {
          React.cloneElement(
            children,
            {
              // @ts-ignore
              onSave: (e: unknown) => {
                // @ts-ignore
                children.props.onSave?.(e);
                save();
                modalRef.current?.closeModal();
              },
            },
          )
        }
      </div>
    </Modal>
  );
}
