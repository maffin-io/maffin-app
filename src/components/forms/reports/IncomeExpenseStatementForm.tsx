'use client';

import React from 'react';
import { Interval } from 'luxon';
import { Controller, useForm } from 'react-hook-form';

import DateRangeInput from '@/components/DateRangeInput';
import { useInterval } from '@/hooks/state';
import IncomeExpenseStatement from '@/templates/pdf/IncomeExpenseStatement';
import { useAccounts, useIncomeStatement } from '@/hooks/api';
import type { Account } from '@/book/entities';
import type { AccountsTotals } from '@/types/book';

type FormValues = {
  interval?: Interval;
};

export default function IncomeExpenseStatementForm(): React.JSX.Element {
  const { data: interval } = useInterval();
  const form = useForm<FormValues>({
    defaultValues: {
      interval,
    },
  });

  const i = form.watch('interval');

  const { data: accounts } = useAccounts();
  const { data: totals } = useIncomeStatement(i);

  return (
    <form
      onSubmit={form.handleSubmit(data => onSubmit(
        data,
        accounts as Account[],
        totals as AccountsTotals,
      ))}
    >
      <div className="flex flex-col items-center text-sm gap-2 pt-5">
        <fieldset>
          <label htmlFor="intervalInput" className="inline-block text-sm my-2">Select dates</label>
          <Controller
            control={form.control}
            name="interval"
            render={({ field }) => (
              <DateRangeInput
                id="intervalInput"
                interval={interval}
                onChange={field.onChange}
              />
            )}
          />
        </fieldset>
      </div>

      <div className="flex w-full justify-center mt-5">
        <button
          className="btn btn-primary capitalize"
          type="submit"
        >
          Submit
        </button>
      </div>
    </form>
  );
}

async function onSubmit(data: FormValues, accounts: Account[], totals: AccountsTotals) {
  const { pdf } = await import('@react-pdf/renderer');
  const myDoc = await IncomeExpenseStatement({
    interval: data.interval as Interval,
    accounts,
    totals,
  });
  const blob = await pdf(myDoc).toBlob();
  window.open(URL.createObjectURL(blob));
}
