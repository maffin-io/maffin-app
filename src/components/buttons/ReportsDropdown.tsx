import React from 'react';
import { BiBarChartAlt, BiNotepad } from 'react-icons/bi';
import IncomeExpenseStatementForm from '../forms/reports/IncomeExpenseStatementForm';
import FormButton from './FormButton';

export default function ReportsDropdown(): JSX.Element {
  return (
    <div className="group relative h-full">
      <button
        type="button"
        className="group-hover:bg-cyan-700/80 dark:group-hover:bg-cyan-600 flex h-full w-full items-center btn btn-primary"
      >
        <BiNotepad className="mr-1" />
        Reports
      </button>
      <ul className="absolute rounded-md w-40 hidden py-2 group-hover:block bg-background-800">
        <li className="text-sm hover:bg-background-700">
          <FormButton
            id="IE report"
            className="flex items-center text-left px-3 py-2 w-full text-cyan-700 hover:text-cyan-600 whitespace-nowrap"
            modalTitle="Income statement"
            buttonContent={(
              <>
                <BiBarChartAlt className="mr-1" />
                Income Statement
              </>
            )}
          >
            <IncomeExpenseStatementForm />
          </FormButton>
        </li>
      </ul>
    </div>
  );
}
