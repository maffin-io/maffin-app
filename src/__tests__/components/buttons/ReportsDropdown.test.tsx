import React from 'react';
import {
  render,
  screen,
} from '@testing-library/react';

import IncomeExpenseStatementForm from '@/components/forms/reports/IncomeExpenseStatementForm';
import ReportsDropdown from '@/components/buttons/ReportsDropdown';
import FormButton from '@/components/buttons/FormButton';

jest.mock('@/components/buttons/FormButton', () => jest.fn(
  (props: React.PropsWithChildren) => (
    <div data-testid="FormButton">
      {props.children}
    </div>
  ),
));

jest.mock('@/components/forms/reports/IncomeExpenseStatementForm', () => jest.fn(
  () => <div data-testid="IncomeExpenseStatementForm" />,
));

describe('ReportsDropdown', () => {
  beforeEach(async () => {
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders as expected', async () => {
    const { container } = render(<ReportsDropdown />);

    await screen.findAllByTestId('FormButton');
    expect(FormButton).toHaveBeenCalledTimes(1);
    expect(FormButton).toHaveBeenNthCalledWith(
      1,
      expect.objectContaining({
        id: 'IE report',
        modalTitle: 'Income statement',
      }),
      undefined,
    );
    expect(IncomeExpenseStatementForm).toHaveBeenCalledTimes(1);
    expect(IncomeExpenseStatementForm).toHaveBeenNthCalledWith(1, {}, undefined);

    expect(container).toMatchSnapshot();
  });
});
