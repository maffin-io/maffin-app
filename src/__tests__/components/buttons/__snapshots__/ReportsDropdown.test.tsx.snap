// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReportsDropdown renders as expected 1`] = `
<div>
  <div
    class="group relative h-full"
  >
    <button
      class="group-hover:bg-cyan-700/80 dark:group-hover:bg-cyan-600 flex h-full w-full items-center btn btn-primary"
      type="button"
    >
      <svg
        class="mr-1"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M19 4h-3V2h-2v2h-4V2H8v2H5c-1.103 0-2 .897-2 2v14c0 1.103.897 2 2 2h14c1.103 0 2-.897 2-2V6c0-1.103-.897-2-2-2zM5 20V7h14V6l.002 14H5z"
        />
        <path
          d="M7 9h10v2H7zm0 4h5v2H7z"
        />
      </svg>
      Reports
    </button>
    <ul
      class="absolute rounded-md w-40 hidden py-2 group-hover:block bg-background-800"
    >
      <li
        class="text-sm hover:bg-background-700"
      >
        <div
          data-testid="FormButton"
        >
          <div
            data-testid="IncomeExpenseStatementForm"
          />
        </div>
      </li>
    </ul>
  </div>
</div>
`;
